src/csharp/Extensions/Microsoft.Spark.Extensions.FileSystem/FileSystem.cs

```diff
@@ -0,0 +1,51 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using Microsoft.Spark.Interop;
+using Microsoft.Spark.Interop.Ipc;
+
+namespace Microsoft.Spark.Extensions.FileSystem
+{
+    /// <summary>
+    /// An abstract base class for a fairly generic filesystem. It may be implemented as a distributed filesystem, or
+    /// as a "local" one that reflects the locally-connected disk. The local version exists for small Hadoop instances
+    /// and for testing.
+    /// 
+    /// All user code that may potentially use the Hadoop Distributed File System should be written to use an FileSystem
+    /// object. The Hadoop DFS is a multi-machine system that appears as a single disk.It's useful because of its fault
+    /// tolerance and potentially very large capacity.
+    /// </summary>
+    public abstract class FileSystem : IDisposable
+    {
+        /// <summary>
+        /// Returns the configured FileSystem implementation.
+        /// </summary>
+        /// <param name="sparkContext">The SparkContext whose configuration will be used.</param>
+        /// <returns>The FileSystem.</returns>
+        public static FileSystem Get(SparkContext sparkContext)
```

Comment:
This is weird. Is this like a factory? How can I create a new type of FileSystem?

Why not just expose the Hadoop FileSystem directly?

Reply:
This is the pattern from the Hadoop `FileSystem` class:  https://hadoop.apache.org/docs/current/api/org/apache/hadoop/fs/FileSystem.html#get-org.apache.hadoop.conf.Configuration-

FileSystem` is an abstract class with static `get()` factory methods that return concrete implementations based on the configuration parameters.

For my .NET implementation, I've added an override of `Get()` that takes `SparkContext` so that we don't have to expose `SparkContext.hadoopConfiguration`.

Reply:
If you are mimicking the hadoop `FileSystem`, shall we follow the same signature (expose minimal `Configuration`)? Also, we should add `Hadoop` to the namespace? (and let's add the link to the comment as well)

Reply:
Added `Hadoop` to the namespace.

For this PR, I just wanted to provide an MVP skeleton so that it would be easy for community members to contribute APIs in additional PRs.  I'm thinking that we can invite others to contribute `Configuration` if they want to.  Thoughts?

Reply:
Looks like we cannot define `Configuration` class in the extension package, since `SparkContext.hadoopConfiguration` will be inside `Microsoft.Spark`..

What if we add `Hadoop` directory under https://github.com/dotnet/spark/tree/master/src/csharp/Microsoft.Spark and add `FileSystem.cs` and `Configuration.cs`? Note that we don't have to expose any of the APIs for `Configuration`.  I just want to be able to create `FileSystem` by `FileSystem.Get(sparkContext.HadoopConfiguration)`. Since we are at 1.0, we want to avoid breaking public APIs if possible.

Reply:
Thoughts @rapoth? I know you wanted to put `FileSystem` in an extension.

Reply:
Synced with @rapoth.  We will go with @imback82 's approach.